File: bzz-spark/src/main/java/com/spark/study/TransformationOperation.java
package com.spark.study;

import org.apache.spark.SparkConf;
import org.apache.spark.SparkContext;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.*;
import scala.Tuple2;

import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

public class TransformationOperation {
	
	
	public static void map(){
		SparkConf conf = new SparkConf().setMaster("local").setAppName("TransformationOperation");
		SparkContext sc = new SparkContext(conf);
		
		List<Integer> numbers = Arrays.asList(1,2,3,4,5);
		
		JavaSparkContext jsc = new JavaSparkContext(sc);
		JavaRDD<Integer> numbersRDD = jsc.parallelize(numbers, 4);
		JavaRDD<Integer> numbersRDDMap = numbersRDD.map(new Function<Integer, Integer>() {
			public Integer call(Integer v1) throws Exception {
				return v1 * 2;
			}
		});
		numbersRDDMap.foreach(new VoidFunction<Integer>() {
			public void call(Integer v1) throws Exception {
				System.out.println(v1 + ",");
			}
		});
		jsc.close();
	}
	
	public static void filter(){
		SparkConf conf = new SparkConf().setMaster("local").setAppName("TransformationOperation");
		SparkContext sc = new SparkContext(conf);
		
		List<Integer> numbers = Arrays.asList(1,2,3,4,5);
		
		JavaSparkContext jsc = new JavaSparkContext(sc);
		
		JavaRDD<Integer> numbersRDD = jsc.parallelize(numbers, 4);
		JavaRDD<Integer> numbersRDDMap = numbersRDD.filter(new Function<Integer, Boolean>() {
			public Boolean call(Integer v1) throws Exception {
				if (v1 % 2 == 0) {
					return false;
				}
				return true;
			}
		});
		
		numbersRDDMap.foreach(new VoidFunction<Integer>() {
			public void call(Integer v1) throws Exception {
				System.out.println(v1 + ",");
			}
		});
		jsc.close();
	}
	
	public static void flatMap(){
		SparkConf conf = new SparkConf().setMaster("local").setAppName("flatMap");
		JavaSparkContext jsc = new JavaSparkContext(conf);
		List<String> words = Arrays.asList("hello you hello me hello world");
		JavaRDD<String> lines = jsc.parallelize(words);
		JavaRDD<String> stringJavaRDD = lines.flatMap(new FlatMapFunction<String, String>() {
			
			public Iterator<String> call(String s) throws Exception {
				String[] split = s.split(" ");
				List<String> strings = Arrays.asList(split);
				return strings.iterator();
			}
		});
		stringJavaRDD.foreach(new VoidFunction<String>() {
			public void call(String s) throws Exception {
				System.out.println(s);
				
			}
		});
		jsc.close();
	}
	
	
	public static void groupByKey(){
		SparkConf conf = new SparkConf().setMaster("local").setAppName("groupByKey");
		JavaSparkContext jsc = new JavaSparkContext(conf);
		
		List<Tuple2<String,Integer>> scores = Arrays.asList(
				new Tuple2<String, Integer>("math", 80),
				new Tuple2<String, Integer>("english", 81),
				new Tuple2<String, Integer>("yuwen", 90),
				new Tuple2<String, Integer>("math", 85),
				new Tuple2<String, Integer>("english", 65),
				new Tuple2<String, Integer>("yuwen", 26)
		);
		
		JavaPairRDD<String, Integer> scoresRDD = jsc.parallelizePairs(scores);
		JavaPairRDD<String, Iterable<Integer>> scoresIterable = scoresRDD.groupByKey();
		scoresIterable.foreach(new VoidFunction<Tuple2<String, Iterable<Integer>>>() {
			public void call(Tuple2<String, Iterable<Integer>> t) throws Exception {
				
				String str = "";
				for(Iterator<Integer> iterator = t._2.iterator();iterator.hasNext();){
					str += iterator.next()+",";
				}
				String substring = str.substring(0, str.length() - 1);
				System.out.println(t._1+" ==> "+substring);
				
			}
		});
		jsc.close();
	}
	public static void reduceByKey(){
		SparkConf conf = new SparkConf().setMaster("local").setAppName("groupByKey");
		JavaSparkContext jsc = new JavaSparkContext(conf);
		
		List<Tuple2<String,Integer>> scores = Arrays.asList(
				new Tuple2<String, Integer>("math", 80),
				new Tuple2<String, Integer>("english", 81),
				new Tuple2<String, Integer>("yuwen", 90),
				new Tuple2<String, Integer>("math", 85),
				new Tuple2<String, Integer>("english", 65),
				new Tuple2<String, Integer>("yuwen", 26)
		);
		JavaPairRDD<String, Integer> scoresRDD = jsc.parallelizePairs(scores);
		JavaPairRDD<String, Integer> totalRDD = scoresRDD.reduceByKey(new Function2<Integer, Integer, Integer>() {
			public Integer call(Integer v1, Integer v2) throws Exception {
				return v1 + v2;
			}
		});
		totalRDD.foreach(new VoidFunction<Tuple2<String, Integer>>() {
			public void call(Tuple2<String, Integer> t) throws Exception {
				
				System.out.println(t._1 +":" + t._2);
			}
		});
		jsc.close();
	}
	
public static void sortByKey(){
		SparkConf conf = new SparkConf().setMaster("local").setAppName("groupByKey");
		JavaSparkContext jsc = new JavaSparkContext(conf);
		
		List<Tuple2<String,Integer>> scores = Arrays.asList(
				new Tuple2<String, Integer>("math", 80),
				new Tuple2<String, Integer>("english", 81),
				new Tuple2<String, Integer>("yuwen", 90),
				new Tuple2<String, Integer>("math", 85),
				new Tuple2<String, Integer>("english", 65),
				new Tuple2<String, Integer>("yuwen", 26)
		);
		JavaPairRDD<String, Integer> scoresRDD = jsc.parallelizePairs(scores);
		
		JavaPairRDD<Integer, String> sorts = scoresRDD.mapToPair(new PairFunction<Tuple2<String, Integer>, Integer, String>() {
			public Tuple2<Integer, String> call(Tuple2<String, Integer> v) throws Exception {
				return new Tuple2<Integer, String>(v._2, v._1);
			}
		});
		JavaPairRDD<Integer, String> sorces = sorts.sortByKey();
		sorces.foreach(new VoidFunction<Tuple2<Integer, String>>() {
			public void call(Tuple2<Integer, String> t) throws Exception {
				System.out.println(t._2 +":" + t._1);
			}
		});
		jsc.close();
	}
	public static void join(){
		SparkConf conf = new SparkConf().setMaster("local").setAppName("groupByKey");
		JavaSparkContext jsc = new JavaSparkContext(conf);
		
		List<Tuple2<Integer,String>> students = Arrays.asList(
				new Tuple2<Integer,String>(1,"math"),
				new Tuple2<Integer,String>(2,"english"),
				new Tuple2<Integer,String>(3,"yuwen")
		);
		List<Tuple2<Integer,Integer>> scores = Arrays.asList(
				new Tuple2<Integer,Integer>(1,80),
				new Tuple2<Integer,Integer>(2,90),
				new Tuple2<Integer,Integer>(3,70)
		);
		
		JavaPairRDD<Integer, String> studentsRDD = jsc.parallelizePairs(students);
		JavaPairRDD<Integer, Integer> scoresRDD = jsc.parallelizePairs(scores);
		JavaPairRDD<Integer, Tuple2<String, Integer>> sorcesJoin = studentsRDD.join(scoresRDD);

		sorcesJoin.foreach(new VoidFunction<Tuple2<Integer, Tuple2<String, Integer>>>() {
			public void call(Tuple2<Integer, Tuple2<String, Integer>> v) throws Exception {
				System.out.println(v._1+" ===> "+v._2._1+" ===> " + v._2._2);
			}
		});
		jsc.close();
	}
	public static void cogroup(){
		SparkConf conf = new SparkConf().setMaster("local").setAppName("groupByKey");
		JavaSparkContext jsc = new JavaSparkContext(conf);
		
		List<Tuple2<Integer,String>> students = Arrays.asList(
				new Tuple2<Integer,String>(1,"math"),
				new Tuple2<Integer,String>(2,"english"),
				new Tuple2<Integer,String>(3,"yuwen")
		);
		List<Tuple2<Integer,Integer>> scores = Arrays.asList(
				new Tuple2<Integer,Integer>(1,80),
				new Tuple2<Integer,Integer>(2,90),
				new Tuple2<Integer,Integer>(3,70)
		);
		
		JavaPairRDD<Integer, String> studentsRDD = jsc.parallelizePairs(students);
		JavaPairRDD<Integer, Integer> scoresRDD = jsc.parallelizePairs(scores);
		JavaPairRDD<Integer, Tuple2<Iterable<String>, Iterable<Integer>>> cogroupRDD = studentsRDD.cogroup(scoresRDD);
		
		cogroupRDD.foreach(new VoidFunction<Tuple2<Integer, Tuple2<Iterable<String>, Iterable<Integer>>>>() {
			public void call(Tuple2<Integer, Tuple2<Iterable<String>, Iterable<Integer>>> t) throws Exception {
				System.out.println(t._1+":"+t._2._1.iterator().next()+":"+t._2._2.iterator().next());
			}
		});
		jsc.close();
	}
	
	public static void main(String[] args) {
		//map();
		//filter();
		
		//flatMap();
		//groupByKey();
		//reduceByKey();
		/*sortByKey();*/
		//join();
		/*cogroup();*/
	}
}

File: bzz-spark/src/main/scala/com/spark/scala/study/TransformationOperation.scala
package com.spark.scala.study

import org.apache.spark.{SparkConf, SparkContext}

object TransformationOperation {



  def map(): Unit ={
    val sparkConf = new SparkConf().setMaster("local").setAppName("TransformationOperation")

    val sparkContext = new SparkContext(config = sparkConf)

    val nums : List[Int] = List(1,2,3,4,5)

    val sparkRDD = sparkContext.parallelize(nums)
    val numberRDD =  sparkRDD.map(num => num * 2)

    numberRDD.foreach(num => {
      println(num )
    })
  }

  def filter():Unit = {
    val sparkConf = new SparkConf().setMaster("local").setAppName("TransformationOperation")
    val sparkContext = new SparkContext(sparkConf)
    val nums : List[Int] = List(1,2,3,4,5)
    val numbersRDD = sparkContext.parallelize(nums)
    val numRDD = numbersRDD.filter(num => num % 2 == 1)
    numRDD.foreach(num =>{
      println(num)
    })
  }
  def flatMap():Unit = {
    val sparkConf = new SparkConf().setMaster("local").setAppName("flatMap")
    val sparkContext = new SparkContext(sparkConf)
    val nums : List[String] = List("hello you ","hello me ","hello world ","hello she ","hello he ")
    val numsRDD = sparkContext.parallelize(nums)
    val wordsRDD = numsRDD.flatMap(num => num.split(" "))
    wordsRDD.foreach(s => println(s))
  }
  def groupByKey():Unit = {
    val conf = new SparkConf().setMaster("local").setAppName("groupByKey");
    val context = new SparkContext(conf)

    val socres : List[Tuple2[String,Int]] = List(
      ("math1",80),
      ("math2",80),
      ("math3",80),
      ("math1",81),
      ("math2",65),
      ("math3",90)
    )
    val scoresPairRDD = context.parallelize(socres);
    val scoresRDDIterate = scoresPairRDD.groupByKey()
    scoresRDDIterate.foreach(score => score._2.foreach(t => println(score._1+" ===> "+ t)))

  }

  def reduceByKey():Unit = {
    val conf = new SparkConf().setMaster("local").setAppName("groupByKey");
    val context = new SparkContext(conf)

    val socres : List[Tuple2[String,Int]] = List(
      ("math1",80),
      ("math2",80),
      ("math3",80),
      ("math1",81),
      ("math2",65),
      ("math3",90)
    )
    val scoresPairRDD = context.parallelize(socres);
    val scoresRDDIterate = scoresPairRDD.reduceByKey(_ + _)
    scoresRDDIterate.foreach(score => println(score._1+" ===> " + score._2))
  }
  def sortByKey():Unit = {
    val conf = new SparkConf().setMaster("local").setAppName("groupByKey");
    val context = new SparkContext(conf)

    val socres : List[Tuple2[String,Int]] = List(
      ("math1",80),
      ("math2",80),
      ("math3",80),
      ("math1",81),
      ("math2",65),
      ("math3",90)
    )
    val scoresPairRDD = context.parallelize(socres);
    val scoresPairRDDs = scoresPairRDD.map(x => (x._2,x._1))

    val scoresRDDIterate = scoresPairRDDs.sortByKey()
    scoresRDDIterate.foreach(score => println(score._2+" ===> " + score._1))
  }
  def join():Unit = {
    val conf = new SparkConf().setMaster("local").setAppName("groupByKey");
    val context = new SparkContext(conf)
    val student : List[Tuple2[Int,String]] = List(
      (1,"math1"),
      (2,"math2"),
      (3,"math3")
    )
    val socres : List[Tuple2[Int,Int]] = List(
      (1,80),
      (2,85),
      (3,82)
    )
    val studentPairRDD = context.parallelize(student);
    val scoresPairRDD = context.parallelize(socres);
    val ssPairRDD = studentPairRDD.join(scoresPairRDD).sortBy(x => x._2._2)

    ssPairRDD.foreach(score => println(score._1 + " ===> " + score._2._1 + " ===> " + score._2._2))
  }
  def cogroup():Unit = {
    val conf = new SparkConf().setMaster("local").setAppName("groupByKey");
    val context = new SparkContext(conf)
    val student : List[Tuple2[Int,String]] = List(
      (1,"math1"),
      (2,"math2"),
      (3,"math3")
    )
    val socres : List[Tuple2[Int,Int]] = List(
      (1,80),
      (2,85),
      (3,82)
    )
    val studentPairRDD = context.parallelize(student);
    val scoresPairRDD = context.parallelize(socres);
    val ssPairRDD = studentPairRDD.cogroup(scoresPairRDD)

    ssPairRDD.foreach(score => println(score._1 + " ===> " + score._2._1.iterator.next() + " ===> " + score._2._2.iterator.next()))
  }

  def main(args: Array[String]): Unit = {
    //map()
    //filter()
    //flatMap()
    //groupByKey()
    //reduceByKey()
    //sortByKey()
    //join()
    cogroup()
  }
}
